### DISCUSSION TITLE
How to send HTTP requests for multiple clients without blocking?

### DISCUSSION THREAD

**Discussion by etale-cohomology** [miguelgrinberg/Flask-SocketIO#2103] [General]
How can I send HTTP requests asynchronously for multiple clients, in non-blocking manner? (Ie. how to "yield" HTTP requests?)

I have a Flask websocket route that sends an HTTP request to an external API.  

```python
app = flask.Flask(__name__)
ws  = flask_socketio.SocketIO(async_mode='gevent_uwsgi')

@ws.on('something')
def ws_something(msg):
   // external http request goes here
```

As the HTTP request is returning data, I `emit()` the data back to the client (the HTTP request takes a long time; say, about a minute).
While the HTTP request is returning data, the `emit/sleep` trick works well to "yield", and I can have multiple clients talking to a single uwsgi server and the requests appears to be served in parallel.

However, the **start** of the HTTP request is still blocking.
First I tried `pycurl`, `curl CLI`, and other libraries, but I guess they're not async-friendly, or "yield-able", so now I'm using [geventhttpclient](https://github.com/geventhttpclient/geventhttpclient) assuming it's better, but the initial HTTP request is still blocking, even with `gevent.sleep(0)`.

The result is that, if I have 2 websocket clients (client `A` and client `B`), and client `A` sends a request, and then client `B` sends a request while `uwsgi` is still returning the response to client `A`, client `A`'s response halt/blocks when Flask starts client `B`'s external HTTP request and then resumes (and this is a very noticeable hiccup).

The external HTTP request looks something like:

```python
url = geventhttpclient.URL('http://some-api.com')
gevent.sleep(0)  # spamming `gevent.sleep(0)` for good measure

client = geventhttpclient.HTTPClient.from_url(url, headers=headers)
gevent.sleep(0)  # spamming `gevent.sleep(0)` for good measure

response = client.post(url.request_uri, data)
gevent.sleep(0)  # spamming `gevent.sleep(0)` for good measure

data = response.read(0x400)
while data:
	chunk = data.decode('utf-8')
	flask_socketio.emit('something', {'txt':chunk})
	gevent.sleep(0)  // this loop works as desired
	data = response.read(CHUNK_BDIM)
client.close()
```

[Following this](https://flask-socketio.readthedocs.io/en/latest/deployment.html) I'm running uwsgi like so:

```
uwsgi --http :8001 --master --wsgi-file app.py --callable app --gevent 100 --http-websockets
```

**Comment by miguelgrinberg** [comment]
Not familiar with this geventhttpclient package that you are using. For any non-gevent client you will need to monkey patch the standard library. Have you tried that? https://uwsgi-docs.readthedocs.io/en/latest/Gevent.html#monkey-patching

**Reply by etale-cohomology** [reply]
Hi Miguel, thank you for the reminder.
I had tried to monkey patch before but it had failed

```
  File "/home/da/py38/lib/python3.8/site-packages/trio/_core/_io_epoll.py", line 189, in EpollIOManager
    _epoll = attr.ib(factory=select.epoll)
AttributeError: module 'select' has no attribute 'epoll'
```

I think because of a dependency on `httpx`, or `httpcore` or `trio` (and the fact that gevent doesn't monkey-patch `epoll`?), or something.

But now I don't have that dependency anymore, so monkey patching works.
And after monkey patching I am getting the expected result.

I was hoping to find a monkeypatch-free solution though, thinking `gevent` or some `gevent`-friendly library might offer a solution.
